Call ProcessNewBlock() asynchronously

TheBlueMatt opened the pull request:
This is a bit of a rework of #16175, which is a rework of #12934. Built on top of #16279.

This is a roughly-minimal patchset to get PNB processing blocks in the background in a clean way, however it doesn't *actually* enable any material parallelism yet - it only sets up a structure for significantly untangling cs_main in coming work.

Next steps:
 * Notably, ProcessMessages will now call PNB, get a future for the block being fully processed, and then immediately block on cs_main waiting to process reject messages and other CNodeState data. This is rather easy to resolve using the new CPeerState (see comment about how we should slowly move things into CPeerState to reduce cs_main contention).
 * In parallel, PNB (and, thus, CheckBlock, and the pre-store AcceptBlock bits) should start losing cs_main, this is primarily doable by splitting mapBlockIndex out of cs_main (probably via something like mempool.cs, where a cs_blockindex acts as a read mutex while writers take cs_main. Given the prevalence of cs_main calls just to look up entries in mapBlockIndex, this should improve things pretty significantly.

TheBlueMatt committed:
Add a new peer state tracking class to reduce cs_main contention.

CNodeState was added for validation-state-tracking, and thus,
logically, was protected by cs_main. However, as it has grown to
include non-validation state (taking state from CNode), and as
we've reduced cs_main usage for other unrelated things, CNodeState
is left with lots of cs_main locking in net_processing.

In order to ease transition to something new, this adds only a
dummy CPeerState which is held as a reference for the duration of
message processing.

Note that moving things is somewhat tricky pre validation-thread as
a consistent lockorder must be kept - we can't take a lock on the
new cs_peerstate in anything that's called directly from
validation.

TheBlueMatt committed:
Move BlockChecked to a background thread

As reject messages are required to go out in-order (ie before any
further messages are processed), this sadly requires that we
further delay re-enabling a peer after a block has been processed
by waiting for current validationinterface callbacks to drain.

This commit enables further reduction of cs_main in net_processing
by allowing us to lock cs_peerstate before cs_main in BlockChecked
(ie allows us to move things which are accessed in BlockChecked,
including DoS state and rejects into CPeerState and out of
CNodeState).

TheBlueMatt committed:
Add a callback to indicate a block has been processed

This resolves the performance regression introduced in the previous
commit by always waking the message processing thread after each
block future resolves.

Sadly, this is somewhat awkward - all other validationinterface
callbacks represent an actual change to the global validation state,
whereas this callback indicates only that a call which one
validation "client" made has completed. After going back and forth
for some time I didn't see a materially better way to resolve this
issue, and luckily its a rather simple change, but its far from
ideal. Note that because we absolutely do not want to ever block on
a ProcessNewBlock-returned-future, the callback approach is
critical.

TheBlueMatt committed:
Remove unnecessary cs_mains in denialofservice_tests

9fdf05d70cac4a62d1aeeb4299e2c3a9a866f8af resolved some lock
inversion warnings in denialofservice_tests, but left in a number
of cs_main locks that are unnecessary (introducing lock inversion
warnings in future changes).

TheBlueMatt committed:
Return the AcceptBlock CValidationState directly in ProcessNewBlock

In practice this means that CheckBlock+ContextualCheckBlock are
called with a passed-in CValidationState before we move onto
connecting the best chain. This makes conceptual sense as these
calls represent the DoS checks on a block (ie PoW and malleability)
which the caller almost certainly wants to know about right away
and shouldn't have to wait on a callback for (and other
validationinterface clients shouldn't care about someone submitting
bogus malleated blocks to PNB).

This also makes it much, much easier to move the best chain
activation logic to a background thread as it implies that if PNB
returns with a IsValid() CValidationState we don't need to care
about trying to process (non-malleated) copies of the block from
other peers.

TheBlueMatt committed:
Make ProcessNewBlock return a future instead of an immediate bool

This prepares for making best-chain-activation and disk writes
happen in a separate thread from the caller, even though all
callsites currently block on the return value immediately.

TheBlueMatt committed:
Move net_processing's ProcessNewBlock calls to resolve async.

Essentially, our goal is to not process anything for the given peer
until the block finishes processing (emulating the previous behavior)
without actually blocking the ProcessMessages loops. Obviously, in
most cases, we'll just go on to the next peer and immediately hit a
cs_main lock, blocking us anyway, but this we can slowly improve
that state over time by moving things from CNodeState to CPeerState.

TheBlueMatt committed:
Move mapBlockSource to cs_peerstate from cs_main

This technically resolves a race where entries are added to
mapBlockSource before we know that they're non-malleated and then
removed only after PNB returns, though in practice this wasn't an
issue since all access to mapBlockSource already held cs_peerstate.

TheBlueMatt committed:
Split AcceptBlock into three stages to write to disk in background

To keep the API the same (and for simplicity of clients, ie
net_processing), this splits AcceptBlock into the do-I-want-this
stage, the checking stage, and the writing stage.

ProcessNewBlock calls the do-I-want-this and checking (ie
malleability checking) stuff, and then dumps blocks that pass
into the background thread. In the background, we re-test the
do-I-want-this logic but skip the checking stuff, before writing
the block to disk and activating the best chain.

practicalswift commented on src/net_processing.cpp:
```diff
@@ -3236,9 +3310,46 @@ bool PeerLogicValidation::SendRejectsAndCheckIfBanned(CNode* pnode, bool enable_
 {
     const CChainParams& chainparams = Params();
+    LOCK(cs_peerstate);
+    CPeerState* peerstate = PeerState(pfrom->GetId());
```
Can `PeerState(pfrom->GetId())` return `nullptr` here?

practicalswift commented on src/net_processing.cpp:
```diff
@@ -3531,6 +3653,13 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
         const CNetMsgMaker msgMaker(pto->GetSendVersion());
 
+        LOCK(cs_peerstate);
+        CPeerState* peerstate = PeerState(pto->GetId());
```
Can `PeerState(pfrom->GetId())` return `nullptr` here?

practicalswift commented on src/validation.h:
```diff
@@ -199,22 +200,34 @@ static const uint64_t MIN_DISK_SPACE_FOR_BLOCK_FILES = 550 * 1024 * 1024;
  * @param[in]   fForceProcessing Process this block even if unrequested; used for non-network block sources and whitelisted peers.
- * @param[out]  fNewBlock A boolean which is set to indicate if the block was first received via this call
- * @return True if state.IsValid()
+ * @return      A future which complets with a boolean which is set to indicate if the block was first received via this call
```
Should be "completes" :-)

laanwj commented:
This is a huge structural change. Please, let's discuss concepts and concept ACK here first before jumping to small nits on the current implementation.

TheBlueMatt commented:
Working on a big structural cleanup, will reopen something new once I have a new PR series.

TheBlueMatt committed:
Run the ActivateBestChain in ProcessNewBlock in a background thread

Spawn a background thread at startup which validates each block as
it comes in from ProcessNewBlock, taking advantage of the new
std::future return value to keep tests simple (and the new
net_processing handling of such values async already).

This makes introducing subtle validationinterface deadlocks much
harder as any locks held going into ProcessNewBlock do not interact
with (in the form of lockorder restrictions) locks taken in
validationinterface callbacks.

Note that after this commit, feature_block and feature_assumevalid
tests time out due to increased latency between block processing
when those blocks do not represent a new best block. This will be
resolved in the next commit.

TheBlueMatt commented:
Actually I was somewhat confused by some existing bugs triggering on travis. I think this is good as-is, original description applies.

promag commented on src/rpc/mining.cpp:
```diff
@@ -132,7 +132,9 @@ static UniValue generateBlocks(const CScript& coinbase_script, int nGenerate, ui
-        if (!ProcessNewBlock(Params(), shared_pblock, true, nullptr))
+        CValidationState state;
+        ProcessNewBlock(Params(), shared_pblock, state, true).wait();
+        if (ChainActive().Tip()->GetBlockHash() != pblock->GetHash())
```
A wild block could be mined in between?

TheBlueMatt replied:
Right, but most likely something was mined before the new block got accepted, which is the right error for that. That said, I think there's a bug here that ChainActive().Tip() needs cs_main, which isn't held here.

promag reviewed (commented):
Concept ACK, at least the idea to separate net and validation sounds reasonable to me.

ryanofsky commented:
Marking up for grabs

http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-12.html#l-394

> \<BlueMatt> #16323 and #16324 are up for grabs if anyone wants to work on them, but there seems to be ~zero interest in reviewing them, cause they have wonderfully scary titles (despite the code actually being pretty simple) :)

Also

http://www.erisian.com.au/bitcoin-core-dev/log-2019-11-13.html#l-30

> \<BlueMatt> jnewbery: note that "Add new peer state tracking class to reduce cs_main contention" only vaguely makes sense on its own - the vast majority of the contension is around cs_main for a reason...witout the rest of the patch to make ProcessNewBlock and the like not take cs_main it is a lot of code movement for not very much performance gain.
